data/scripts.js

```diff
@@ -292,11 +255,289 @@ let BattleScripts = {
 			return false;
 		}
 
+		return true;
+	},
+	trySpreadMoveHit(targets, pokemon, move) {
+		if (!targets.length) {
+			this.attrLastMove('[notarget]');
+			this.add(this.gen >= 5 ? '-fail' : '-notarget', pokemon);
+			return false;
+		}
+		if (targets.length > 1) move.spreadHit = true;
+
+		/** @type {((targets: Pokemon[], pokemon: Pokemon, move: ActiveMove) => (number | boolean | "" | undefined)[] | undefined)[]} */
+		let moveSteps = [
+			// 0. check for semi invulnerability
+			this.tryImmunityEvent,
+
+			// 1. check for type immunity (this is step 2 in gen 7+)
+			this.typeImmunity,
+
+			// 2. run the 'TryHit' event (Protect, Magic Bounce, Volt Absorb, etc.) (this is step 1 in gen 7)
+			this.tryHitEvent,
+
+			// 3. check for powder immunity
+			this.powderImmunity,
+
+			// 4. check for prankster immunity
+			this.pranksterImmunity,
+
+			// 5. check accuracy
+			this.accuracy,
+
+			// 6. break protection effects
+			this.breakProtect,
+
+			// 7. steal positive boosts (Spectral Thief)
+			this.stealBoosts,
+
+			// 8. loop that processes each hit of the move (has its own steps per iteration)
+			this.moveHitLoop,
+
+			// 9. effects that run after secondary effects, such as Color Change
+			this.afterMoveSecondaryEvent,
+		];
+		if (this.gen >= 7) {
+			// Swap step 1 with step 2
+			[moveSteps[1], moveSteps[2]] = [moveSteps[2], moveSteps[1]];
+		}
+
+		this.setActiveMove(move, pokemon, targets[0]);
+		move.zBrokeProtect = false;
+
+		let hitResult = this.singleEvent('PrepareHit', move, {}, targets[0], pokemon, move);
+		if (!hitResult) {
+			if (hitResult === false) {
+				this.add('-fail', pokemon);
+				this.attrLastMove('[still]');
+			}
+			return false;
+		}
+		this.runEvent('PrepareHit', pokemon, targets[0], move);
+
+		if (!this.singleEvent('Try', move, null, pokemon, targets[0], move)) {
+			return false;
+		}
+
+		let finalResult;
+		let hitResults;
```

Comment:
```suggestion
		/** @type {(number | boolean | "" | undefined)[] | undefined} */
		let hitResults;
```

Reply:
If I do this, it complains about `hitResults` potentially being undefined even though I already checked for that. It seems that TS just really hates the `filter` function.

Reply:
Meh, I'll just `// @ts-ignore`.

Reply:
Yes, that is in fact what I said:

> Although it's true, TypeScript can't infer that `filter` is synchronous, so you'll have to explicitly cast it.